Subscriber.onComplete() should not raise an exception

```
[armeria-common-worker-nio-4-1] WARN com.linecorp.armeria.common.stream.DefaultStreamMessageDuplicator - Subscriber.onComplete() should not raise an exception. subscriber: com.linecorp.armeria.client.HttpRequestSubscriber@4a4ba28f
java.lang.AssertionError: null
	at com.linecorp.armeria.internal.common.HttpObjectEncoder.writeData(HttpObjectEncoder.java:67)
	at com.linecorp.armeria.client.HttpRequestSubscriber.write(HttpRequestSubscriber.java:283)
	at com.linecorp.armeria.client.HttpRequestSubscriber.onComplete(HttpRequestSubscriber.java:260)
	at com.linecorp.armeria.common.stream.DefaultStreamMessageDuplicator$StreamMessageProcessor.doUnsubscribe(DefaultStreamMessageDuplicator.java:323)
	at com.linecorp.armeria.common.stream.DefaultStreamMessageDuplicator$StreamMessageProcessor.lambda$unsubscribe$4(DefaultStreamMessageDuplicator.java:308)
	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:479)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/core/build.gradle.kts
+++ b/core/build.gradle.kts
@@ -39,6 +39,7 @@ dependencies {
     testImplementation("com.linecorp.armeria:armeria-junit5:1.1.0")
 
     testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.7.0")
+    testRuntimeOnly("ch.qos.logback:logback-classic:1.2.3")
 }
 
 java {

--- a/core/src/main/java/dev/gihwan/tollgate/core/upstream/DefaultUpstream.java
+++ b/core/src/main/java/dev/gihwan/tollgate/core/upstream/DefaultUpstream.java
@@ -25,6 +25,8 @@
 package dev.gihwan.tollgate.core.upstream;
 
 import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.concurrent.CompletableFuture;
 
 import javax.annotation.Nullable;
@@ -35,8 +37,8 @@ import org.slf4j.LoggerFactory;
 import com.google.common.base.Splitter;
 
 import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.AggregatedHttpRequest;
 import com.linecorp.armeria.common.HttpRequest;
-import com.linecorp.armeria.common.HttpRequestDuplicator;
 import com.linecorp.armeria.common.HttpResponse;
 import com.linecorp.armeria.common.HttpStatus;
 import com.linecorp.armeria.common.RequestHeaders;
@@ -67,37 +69,36 @@ public final class DefaultUpstream implements Upstream {
             return HttpResponse.of(HttpStatus.INTERNAL_SERVER_ERROR);
         }
 
-        final HttpRequest duplicatedReq = duplicateRequest(req, path);
-        return sendRequest(duplicatedReq);
+        return HttpResponse.from(req.aggregate()
+                                    .thenApply(aggregated -> duplicateRequest(aggregated, path))
+                                    .thenApply(this::sendRequest));
     }
 
     @Nullable
     private String buildPath(ServiceRequestContext ctx) {
-        final StringBuilder sb = new StringBuilder();
-        for (String token : PATH_SPLITTER.split(config.endpoint().path())) {
-            sb.append(PATH_SEPARATOR);
-            if (token.startsWith("{") && token.endsWith("}")) {
-                final String pathParamName = token.substring(1, token.length() - 1);
+        final List<String> segments = new ArrayList<>();
+        for (String segment : PATH_SPLITTER.split(config.endpoint().path())) {
+            if (segment.startsWith("{") && segment.endsWith("}")) {
+                final String pathParamName = segment.substring(1, segment.length() - 1);
                 final String pathParamValue = ctx.pathParam(pathParamName);
                 if (pathParamValue == null) {
                     logger.error("Path parameter {} is not found on path {}.", pathParamName, ctx.path());
                     return null;
                 }
-                sb.append(pathParamValue);
+                segments.add(pathParamValue);
             } else {
-                sb.append(token);
+                segments.add(segment);
             }
         }
-        return sb.toString();
+        return String.join("/", segments);
     }
 
-    private HttpRequest duplicateRequest(HttpRequest req, String path) {
-        final HttpRequestDuplicator duplicator = req.toDuplicator();
-        final RequestHeaders newHeaders = req.headers().toBuilder()
-                                             .method(config.endpoint().method())
-                                             .path(path)
-                                             .build();
-        return duplicator.duplicate(newHeaders);
+    private HttpRequest duplicateRequest(AggregatedHttpRequest req, String path) {
+        final RequestHeaders headers = req.headers().toBuilder()
+                                          .method(config.endpoint().method())
+                                          .path(path)
+                                          .build();
+        return HttpRequest.of(headers, req.content(), req.trailers());
     }
 
     private HttpResponse sendRequest(HttpRequest req) {

--- a/core/src/test/java/dev/gihwan/tollgate/core/upstream/DefaultUpstreamTest.java
+++ b/core/src/test/java/dev/gihwan/tollgate/core/upstream/DefaultUpstreamTest.java
@@ -1,0 +1,152 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Gihwan Kim
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package dev.gihwan.tollgate.core.upstream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.logging.LoggingService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import dev.gihwan.tollgate.core.service.ServiceConfig;
+
+class DefaultUpstreamTest {
+
+    private static final AtomicReference<AggregatedHttpRequest> reqCapture = new AtomicReference<>();
+
+    @RegisterExtension
+    static final ServerExtension serviceServer = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.service("/path", (ctx, req) -> HttpResponse.from(req.aggregate()
+                                                                   .thenApply(aggregated -> {
+                                                                       reqCapture.set(aggregated);
+                                                                       return HttpResponse.of(HttpStatus.OK);
+                                                                   })));
+        }
+    };
+
+    @RegisterExtension
+    static final ServerExtension tollgateServer = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.decorator(LoggingService.newDecorator());
+
+            configureGetService(sb);
+            configurePostService(sb);
+            configurePutService(sb);
+            configureDeleteService(sb);
+        }
+
+        private void configureGetService(ServerBuilder sb) {
+            final ServiceConfig service = ServiceConfig.of(serviceServer.httpUri().toString());
+            final UpstreamEndpointConfig endpoint = UpstreamEndpointConfig.of(HttpMethod.GET, "/path");
+            final UpstreamConfig config = UpstreamConfig.of(service, endpoint);
+            sb.service("/get", new DefaultUpstream(config));
+        }
+
+        private void configurePostService(ServerBuilder sb) {
+            final ServiceConfig service = ServiceConfig.of(serviceServer.httpUri().toString());
+            final UpstreamEndpointConfig endpoint = UpstreamEndpointConfig.of(HttpMethod.POST, "/path");
+            final UpstreamConfig config = UpstreamConfig.of(service, endpoint);
+            sb.service("/post", new DefaultUpstream(config));
+        }
+
+        private void configurePutService(ServerBuilder sb) {
+            final ServiceConfig service = ServiceConfig.of(serviceServer.httpUri().toString());
+            final UpstreamEndpointConfig endpoint = UpstreamEndpointConfig.of(HttpMethod.PUT, "/path");
+            final UpstreamConfig config = UpstreamConfig.of(service, endpoint);
+            sb.service("/put", new DefaultUpstream(config));
+        }
+
+        private void configureDeleteService(ServerBuilder sb) {
+            final ServiceConfig service = ServiceConfig.of(serviceServer.httpUri().toString());
+            final UpstreamEndpointConfig endpoint = UpstreamEndpointConfig.of(HttpMethod.DELETE, "/path");
+            final UpstreamConfig config = UpstreamConfig.of(service, endpoint);
+            sb.service("/delete", new DefaultUpstream(config));
+        }
+    };
+
+    @Test
+    void proxyGet() {
+        final WebClient webClient = WebClient.builder(tollgateServer.httpUri()).build();
+
+        final AggregatedHttpResponse res = webClient.get("/get").aggregate().join();
+        assertThat(res.status()).isEqualTo(HttpStatus.OK);
+
+        final AggregatedHttpRequest req = reqCapture.get();
+        assertThat(req.method()).isEqualTo(HttpMethod.GET);
+        assertThat(req.path()).isEqualTo("/path");
+    }
+
+    @Test
+    void proxyPost() {
+        final WebClient webClient = WebClient.builder(tollgateServer.httpUri()).build();
+
+        final AggregatedHttpResponse res = webClient.post("/post", "Hello, World!").aggregate().join();
+        assertThat(res.status()).isEqualTo(HttpStatus.OK);
+
+        final AggregatedHttpRequest req = reqCapture.get();
+        assertThat(req.method()).isEqualTo(HttpMethod.POST);
+        assertThat(req.path()).isEqualTo("/path");
+        assertThat(req.contentUtf8()).isEqualTo("Hello, World!");
+    }
+
+    @Test
+    void proxyPut() {
+        final WebClient webClient = WebClient.builder(tollgateServer.httpUri()).build();
+
+        final AggregatedHttpResponse res = webClient.post("/put", "Hello, World!").aggregate().join();
+        assertThat(res.status()).isEqualTo(HttpStatus.OK);
+
+        final AggregatedHttpRequest req = reqCapture.get();
+        assertThat(req.method()).isEqualTo(HttpMethod.PUT);
+        assertThat(req.path()).isEqualTo("/path");
+        assertThat(req.contentUtf8()).isEqualTo("Hello, World!");
+    }
+
+    @Test
+    void proxyDelete() {
+        final WebClient webClient = WebClient.builder(tollgateServer.httpUri()).build();
+
+        final AggregatedHttpResponse res = webClient.get("/delete").aggregate().join();
+        assertThat(res.status()).isEqualTo(HttpStatus.OK);
+
+        final AggregatedHttpRequest req = reqCapture.get();
+        assertThat(req.method()).isEqualTo(HttpMethod.DELETE);
+        assertThat(req.path()).isEqualTo("/path");
+    }
+}

